### ISSUE TITLE
no-undefined-types does not consider @template definitions with type constraints

### ISSUE THREAD

**Issue by Elberet** [gajus/eslint-plugin-jsdoc#428]
When running in typescript mode (`settings.jsdoc.mode = "typescript"`), template types with type constraints are ignored by the `no-undefined-types` rule:

```js
/**
 * Run callback when hooked method is called.
 *
 * @template {BaseObject} T
 * @param {T} obj - object whose method should be hooked.
 * @param {string} method - method which should be hooked.
 * @param {(sender: T) => void} callback - callback which should
 * be called when the hooked method was invoked.
 */
function registerEvent(obj, method, callback) { ... }
```

In the example above, the type constraint on `T` results in `jsdoc/no-undefined-types` emitting a warning "The type 'T' is undefined". See [the Typescript handbook](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#template) for more examples on typed templates.

**Comment by brettz9** [comment]
Fixed by 5f488214bab30864e74c7d5500c075549a65bb1b

**Comment by brettz9** [comment]
And released with https://github.com/gajus/eslint-plugin-jsdoc/releases/tag/v17.1.2 .

**Comment by Elberet** [comment]
😱 

Now that's what I call response time. 👍

**Comment by snowteamer** [comment]
There is still a false positive when the type parameter is used inside the function:
```js
/**
 * @template T
 * @param {T} arg
 * @returns {[T]}
 */
function example(arg) {
	const result =  /** @type {[T]} */ (new Array());
	result[0] = arg;
	return result;
}
```
`The type 'T' is undefined. (eslint(jsdoc/no-undefined-types) [7, 1]`

**Comment by brettz9** [comment]
@snowteamer : Thank you. Could you please file a new issue? Also if know of docs showing the scope of where the template name can be used?

**Comment by wycats** [comment]
I hit the same issue today!

**Comment by Elberet** [comment]
See also #559 for the template type scope issue.
